### DISCUSSION TITLE
tutorials/platform/android/android_in_app_purchases

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#51] [User-contributed notes]
# tutorials/platform/android/android_in_app_purchases

Godot offers a first-party GodotGooglePlayBilling Android plugin compatible with Godot 4 which uses the Google Play Billing library. Usage: Getting started: Make sure you have enabled and successfu...

https://docs.godotengine.org/en/stable/tutorials/platform/android/android_in_app_purchases.html

**Comment by GuacGameDev** [comment]
Trying to get this to work but I can't find a directory called libs/ after cloning the repository. Am I missing a step on the GitHub section?

**Reply by bearcatjamboree** [reply]
probably easier to just download from the releases page:

https://github.com/godotengine/godot-google-play-billing/releases

I think all you need is GodotGooglePlayBilling.1.2.0.release.aar and GodotGooglePlayBilling.gdap

**Reply by GuacGameDev** [reply]
That worked, thank you! Only issue now is I get this warning from Google:

"App must use Google Play Billing Library version 6.0.1 or later"


I'll scour the web to see if I can find anyone who has changed it to the latest. I'm guessing its tied to the GodotGooglePlayBilling.1.2.0.release.aar file

**Reply by cezbloch** [reply]
The 'libs' directory is not in the 'Godot Google Play Plugin'. Just create the 'libs' directory and copy the .aar to it. Then you need to specify Android SDK path. Create a file called 'local.properties' in the root and add a line to it 'sdk.dir=<your_android_sdk_path>'

**Comment by muhal24** [comment]
The Google Play Billing version on the Godot official website is 5.2.1, but Google accepts 6.0.1 or higher according to its updated policies.
So I need version 6.0.1 or higher.

**Comment by JamesArbaes** [comment]
I didn't manage to get this to work at all. The plugin doesn't seem to be recognized. Could it be because my project is using C# instead Godot script?

**Comment by MTGPROD** [comment]
The 1.3.0 aar release do not work with the newly released v2.0.0 of the plugin. I got a 
```
SCRIPT ERROR: Invalid call. Nonexistent function 'purchase' in base 'JNISingleton'.
at: purchase (res://addons/GodotGooglePlayBilling/BillingClient.gd:129)
```

**Reply by keatane** [reply]
Just seen your message from documentation comments.
I had the same problem and it seems to be an overloading problem, but the error is misleading. 
Just pass the second argument as default:
`payment.purchase("product_id", false)` and you should be fine.

It took me a lot to figure out and now it's working. Also I suggest to take as reference the name of the function connected to the singleton, and not the "generic" definition (e.g. consumePurchase vs consume_purchase)

**Reply by carmen-x** [reply]
If you see this error, then make sure you are using the new "addons" folder in favor of the obsolete "android\plugins" folder.  Also, you will now need to enable the plugin using Project > Project Settings > (Plugins tab) rather than enabling it in the Export template.  It is probably a good idea to create a brand new Export template as well and manually copy over each setting just to ensure you do not get any strange behavior from an old build template.

**Comment by X3Yprojekt** [comment]
Where is the demo for v2.0.0, i want to migrate from v1.3

**Comment by thanhngc171** [comment]
Hi, i was able to use the plugin and connect to Billing and retrieve the products i created on Google Play Console. But i cant seem to test the formatted_price if the device is in a different region.

For example if my device is in Vietnam then the formatted_price should return in VND, if my device is in the US then it should return in USD. Is this something handled automatically or do i have to manually do the conversion?

I have tried using Google Billing Lab and switch over to the US, but the product return is still in VND

**Reply by pysleep3** [reply]
if u dont mind may i know did u use in app purchase or one time product

**Comment by Serimert90** [comment]
I think we can't test in local anymore? Purchases are returning 5 code meaning "please sign apk correctly or developer error". I think we need to create test build, upload to play store and make yourself as tester, test from there, am i wrong? Or godot 4.5 beta6 has problems with apk signing?

**Comment by pysleep3** [comment]
i use oppo a9 2020 android version 11, playstore version 48.8.07, for my internal testing also use opt in email.

my question is eventho i successfully connected to console and make sure my plugin_singleton occur and is_ready = true, why i dont even get callback from plugin for product_detail_response (i know in theory eventho wrong id it still give response but mine not even has that response). for ur info only query_purchases give response else not even responded. tq for reading this i hope u dont mind helping me, ur help really mean a lot for me

**Comment by carmen-x** [comment]
The Godot 4 stable build as of this comment is 4.5.1 and the ONLY version of the GodotGooglePlayBilling library that will work in a production setting with Google Play is version 1.3.0 and a Target SDK of at least 35.  

With this in mind, there were breaking changes in both how Godot 4 plugins work and also the GodotGooglePlayBilling library implementation has changed.

If you are upgrading your app to use the latest version of this library/plugin, the first thing you must do is delete the android\plugins folder and instead create an "addons" folder in the ROOT folder of your project.  Now, download the latest version of the library from Github, unzip it and copy the entire "GodotGooglePlayBilling" folder to the "addons" folder.  At this point, it is worth noting the way plugins work is different now.  There isn't a "gdap" file to describe the plugin anymore, but instead a generic "plugin.cfg".  Spend a couple minutes looking at it so you understand the changes that were made in how paths work when creating a debug or release build.

With that out of the way, open your project in Godot 4.5.1 and then go to Project > Project Settings > (Plugins tab) and enable the GodotGooglePlayBilling plugin.  NOTE: Enabling the plugin is no longer in the Export template settings.

To steer clear of any potential upgrade issues, I also created a brand new Android Export Template and copied over all of my settings from the previous one.

After that is done, if you had a working implementation with Google Play prior to version 1.3.0 of the plugin, then all of that code will need to be reworked since the library implementation is fairly different.  

The strange thing is the old library will work in a debug build and payment processing works fine!  However, when you deploy to production then things go awry and you'll be scratching your head wondering what the heck is wrong.  Well, Google's debug environment isn't the same as their production environment.  Lots of pitfalls here, so good luck!

**Comment by beeptobeep** [comment]
I am using Godot 4.5 with the GodotGooglePlayBilling 3.1.0 Release. 
I was able to make the in-app purchase work with connecting to the Google Play Store. I was also able to buy the product. 
However, I want my product to be used as a consumable product and when i try purchasing 
the product again it says “I already own this item.” ( i dont have the acknowledge functions in because it said i dont need it if i am consuming the purchase. )

Im confused on what to do with the handle_purchase_token function, because it tells me to remove “purchase” from the tracking list. The issues I have is that i dont know how to remove the “purchase” and I dont know what the tracking list is, as well as how to grab it in my code. 

Also for consumable function, func _process_purchase(purchase), it says to “add code to store payment” but I dont know what to put there. 

If anyone could help it would be greatly appreciated.

( Also tip for anyone, the way I made it work is you need to connect your app from the Google Play console to Google Cloud. You have to enable the Google Play Android Developer API.
I dont know if there another way to do it without Google Cloud because Im a beginner. )

**Reply by carmen-x** [reply]
I don't think the call to consume purchase is working correctly if you are seeing this error.

I'd put a breakpoint in the signal handler for onPurchaseUpdated and check all properties for the billingResult, in particular the "purchases" key, "response_code" and any "debug_message".  That will get you to the next step.  You'll have to have your Android test device connected and debug remotely. 

You are correct though, don't worry about acknowledging a purchase if you consume it.

The "add code to store payment" comment means you should make sure the payment was actually successful - that is the purchase state is PURCHASED.  I don't know the logic for your game, but I don't put this code in the app itself but rather on my backend so the game won't function without the app authenticating with my backend that the payment was successful. This prevents hackers from cheesing my game and getting as many free items as they want.  This part might be difficult to understand and implement for a beginner, so I can only wish you luck in finding your way here since I could go on for hours about this subject.  This also means I am not using GodotGooglePlayBilling at this point, but rather interacting directly with the GooglePlayApi via REST calls from my cloud app, in particular GetPurchaseInfo(purchaseToken) and then ConsumePurchase(purchaseToken).

As a beginner though, if you just want to get your app out then don't worry too much about using a backend.  Chances are, it is overengineering and concentrating on other things will get you further.  You can always come back to this later if it's a problem.
